shiny/ui/_chat.py

```diff
@@ -1325,6 +1376,210 @@ async def _send_custom_message(self, handler: str, obj: ClientMessage | None):
+            @reactive.event(lambda: self.messages(format=MISSING), ignore_init=True)
+            async def _():
+                messages = self.messages(format=MISSING)
+
+                if len(messages) == 0:
+                    return
+
+                last_message = messages[-1]
+
+                if last_message.get("role") == "assistant":
+                    await session.bookmark()
+
+        @root_session.bookmark.on_bookmark
+        async def cancel_on_bookmark_client(state: BookmarkState):
+            if resolved_bookmark_id_str in state.values:
+                raise ValueError(
+                    f'Bookmark value with id (`"{resolved_bookmark_id_str}"`) already exists.'
+                )
+
+            with reactive.isolate():
+                state.values[resolved_bookmark_id_str] = await get_state()
+
+        @root_session.bookmark.on_bookmark
+        def cancel_on_bookmark_ui(state: BookmarkState):
+            if resolved_bookmark_id_msgs_str in state.values:
+                raise ValueError(
+                    f'Bookmark value with id (`"{resolved_bookmark_id_msgs_str}"`) already exists.'
+                )
+
+            with reactive.isolate():
+                # This only contains `ui.Chat(messages=)`
+                # This does NOT contain the `chat.ui(messages=)` values.
+                # When restoring, the `chat.ui(messages=)` values will need to be kept
+                # and the `ui.Chat(messages=)` values will need to be reset
+                state.values[resolved_bookmark_id_msgs_str] = self.messages(
+                    format=MISSING
+                )
+
+        ###############
+        # On Restore
+
+        # Attempt to stop the initialization of the `ui.Chat(messages=)` messages
+        self._init_chat.destroy()
+
+        @root_session.bookmark.on_restore
+        async def cancel_on_restore_client(state: RestoreState):
+            if resolved_bookmark_id_str not in state.values:
+                return
+
+            # Retrieve the chat turns from the bookmark state
+            info = state.values[resolved_bookmark_id_str]
+            await set_state(info)
+
+        @root_session.bookmark.on_restore
+        async def cancel_on_restore_ui(state: RestoreState):
+            print("set chat ui")
+            # Do not call `self.clear_messages()` as it will clear the
+            # `chat.ui(messages=)` in addition to the `self.messages()`
+            # (which is not what we want).
+
+            # We always want to keep the `chat.ui(messages=)` values
+            # and `self.messages()` are never initialized due to
+            # calling `self._init_chat.destroy()` above
+
+            msgs: list[Any] = state.values[resolved_bookmark_id_msgs_str]
+            if not msgs:
+                # If no messages, set the default messages
+                for default_message in self._init_chat_messages:
+                    await self.append_message(default_message)
+                return
```

Comment:
Related to https://github.com/posit-dev/py-shiny/pull/1951/files#r2023793758, it kinda feels like you could remove that other line and simplify this?

```suggestion
            if msgs:
                self._init_chat.destroy()
```

Reply:
This would be too late. Most likely the effect would have already executed.

It must be destroyed earlier and then reinstated later.